### PR TITLE
ci: make bazel.debug build and test from a consuming project.

### PR DESCRIPTION
**Pull request by htuch**
This allows us to catch in CI breakages such as those fixed in #732. Also, it allows us to validate
our example and story for other projects that consume Envoy for custom linking of filters keeps
working.

This adds the requirement of bind mounting in docker run with -v <some path>:/build for bazel.debug
and bazel.coverage builds.

### CODE REVIEW

File: ci/build_setup.sh [hunk 1]
```diff
@@ -14,23 +14,58 @@ NUM_CPUS=`grep -c ^processor /proc/cpuinfo`
...
+  if [[ ! -a "${ENVOY_CONSUMER_SRCDIR}" ]]
+  then
+    # TODO(htuch): Update to non-htuch when https://github.com/lyft/envoy/issues/404 is sorted.
+    git clone https://github.com/htuch/envoy-consumer.git "${ENVOY_CONSUMER_SRCDIR}"
+  fi
+  cp -f "${ENVOY_SRCDIR}"/ci/WORKSPACE.consumer "${ENVOY_CONSUMER_SRCDIR}"/WORKSPACE
+  # This is the hash on https://github.com/htuch/envoy-consumer.git we pin to.
+  (cd "${ENVOY_CONSUMER_SRCDIR}" && git checkout 94e11fa753a1e787c82cccaec642eda5e5b61ed8)
+
+  # Also setup some space for building Envoy standadlone.
```
Comment (mattklein123) [inline ci/build_setup.sh]: typo standadlone

File: ci/do_ci.sh [hunk 2]
```diff
@@ -9,29 +9,31 @@ echo "building using ${NUM_CPUS} CPUs"
 
 if [[ "$1" == "bazel.debug" ]]; then
   echo "bazel debug build with tests..."
-  ${BAZEL} build ${BAZEL_BUILD_OPTIONS} //source/exe:envoy-static
-  ${BAZEL} test ${BAZEL_BUILD_OPTIONS} --test_output=all \
-    --cache_test_results=no //test/...
+  cd "${ENVOY_CONSUMER_SRCDIR}"
```
Comment (mattklein123) [inline ci/do_ci.sh]: Is there value in continuing to build/test just the main project? I can't think of anything off the top of my head but I'm wondering if we are going to miss some type of failure that only shows up when attempting to built the primary project standalone?

File: ci/build_setup.sh [hunk 3]
```diff
@@ -14,23 +14,58 @@ NUM_CPUS=`grep -c ^processor /proc/cpuinfo`
 
 if [[ "$1" == bazel* ]]
 then
+  export BUILD_DIR=/build
+  # Make sure that "docker run" has a -v bind mount for /build, since cmake
+  # users will only have a bind mount for /source.
+  if [[ ! -a "${BUILD_DIR}" ]]
```
Comment (mattklein123) [inline ci/build_setup.sh]: does -d work for the bind mounts? IIRC -a is deprecated so maybe use -e ?

File: ci/build_setup.sh [hunk 4]
```diff
@@ -14,23 +14,58 @@ NUM_CPUS=`grep -c ^processor /proc/cpuinfo`
...
-  export TEST_TMPDIR=/source/bazel-build
-  export SRCDIR="$(realpath "${PWD}")"
+  export TEST_TMPDIR=/build/tmp
   export BAZEL="bazel"
   export BAZEL_COVERAGE="bazel-coverage"
   # Not sandboxing, since non-privileged Docker can't do nested namespaces.
-  export BAZEL_BUILD_OPTIONS="--strategy=CppCompile=standalone --strategy=CppLink=standalone \
-    --strategy=TestRunner=standalone --strategy=ProtoCompile=standalone \
-    --strategy=Genrule=standalone --verbose_failures --package_path %workspace%:.."
+  export BAZEL_BUILD_OPTIONS="--strategy=Genrule=standalone --spawn_strategy=standalone \
```
Comment (mattklein123) [inline ci/build_setup.sh]: Why the change in options?

Comment (htuch) [inline ci/build_setup.sh]: Simplified, this is the easiest way to disable all sandboxing on the command-line.

File: ci/do_ci.sh (hunk 2 above)
Comment (htuch) [inline ci/do_ci.sh]: I think we can cover the build/test of just the main project in the coverage run. When we formalize the build/test matrix we can make this explicit. I think it's useful to make sure that we're checking the main project build/test is sane, but it's not worth burning CI on it today at least (when we free up jobs when removing cmake we could consider this).

File: ci/build_setup.sh (hunk 4 above)
Comment (htuch) [inline ci/build_setup.sh]: (and package_path had to change since we've changed the relative build location)

File: ci/do_ci.sh (hunk 2 above)
Comment (mattklein123) [inline ci/do_ci.sh]: Yup makes sense. As long as we do at least one build that doesn't use consumer I think that's fine.

File: ci/build_setup.sh (hunk 4 above)
Comment (mattklein123) [inline ci/build_setup.sh]: When I build on my own machine inside the consuming project, I don't seem to need to set --package_path. Out of curiosity why is this necessary here?

Comment (mattklein123) [inline ci/build_setup.sh]: Maybe because of the submodule situation?

Comment (htuch) [inline ci/build_setup.sh]: We're building for non-consuming now in `/build/envoy`, this is where we invoke Bazel. So, it needs to know that it can still find the Envoy source in `/source` https://bazel.build/versions/master/docs/bazel-user-manual.html#flag--package_path
